core/trie_verify.go

```diff
@@ -0,0 +1,345 @@
+	// load unverified blocks in a normalized chain and start a batch of verify task
+	header := vm.bc.CurrentHeader()
+	// Start verify task from H to H-11 if need.
+	vm.NewBlockVerifyTask(header)
+	pruneTicker := time.NewTicker(pruneInterval)
+	defer pruneTicker.Stop()
+	for {
+		select {
+		case h := <-vm.newTaskCh:
+			vm.NewBlockVerifyTask(h)
+		case hash := <-vm.verifyCh:
+			vm.cacheBlockVerified(hash)
+			if task, ok := vm.tasks[hash]; ok {
+				delete(vm.tasks, hash)
+				close(task.terminalCh)
+			}
+		case <-pruneTicker.C:
+			for hash, task := range vm.tasks {
+				if vm.bc.CurrentHeader().Number.Cmp(task.blockHeader.Number) == 1 &&
+					vm.bc.CurrentHeader().Number.Uint64()-task.blockHeader.Number.Uint64() > pruneHeightDiff {
+					delete(vm.tasks, hash)
+					close(task.terminalCh)
+				}
+			}
+		case message := <-vm.messageCh:
+			if vt, ok := vm.tasks[message.verifyResult.BlockHash]; ok {
+				vt.messageCh <- message
+			}
+		case <-vm.bc.quit:
+			for _, task := range vm.tasks {
+				close(task.terminalCh)
+			}
+			return
+		}
+	}
+}
+
+func (vm *verifyManager) NewBlockVerifyTask(header *types.Header) {
+	for i := 0; header != nil && i <= maxForkHeight; i++ {
+		func(hash common.Hash) {
+			// if verified cache record that this block has been verified, skip.
+			if _, ok := vm.verifiedCache.Get(hash); ok {
+				return
+			}
+			// if there already has a verify task for this block, skip.
+			if _, ok := vm.tasks[hash]; ok {
+				return
+			}
+			diffLayer := vm.bc.GetTrustedDiffLayer(hash)
+			// if this block has no diff, there is no need to verify it.
+			var err error
+			if diffLayer == nil {
+				if diffLayer, err = vm.bc.GenerateDiffLayer(hash); err != nil {
+					log.Error("failed to get diff layer", "block", hash, "number", header.Number, "error", err)
+					return
+				}
+			}
+			diffHash, err := GetTrustedDiffHash(diffLayer)
+			if err != nil {
+				log.Error("failed to get diff hash", "block", hash, "number", header.Number, "error", err)
+				return
+			}
+			verifyTask := NewVerifyTask(diffHash, header, vm.peers, vm.verifyCh, vm.allowUntrusted)
+			vm.tasks[hash] = verifyTask
+		}(header.Hash())
+		header = vm.bc.GetHeaderByHash(header.ParentHash)
+	}
+}
+
+func (vm *verifyManager) cacheBlockVerified(hash common.Hash) {
+	if vm.verifiedCache.Len() >= verifiedCacheSize {
+		vm.verifiedCache.RemoveOldest()
+	}
+	vm.verifiedCache.Add(hash, true)
+}
+
+// AncestorVerified function check block has been verified or it's a empty block.
+func (vm *verifyManager) AncestorVerified(header *types.Header) bool {
+	// find header of H-11 block.
+	header = vm.bc.GetHeaderByNumber(header.Number.Uint64() - maxForkHeight)
+	// If start from genesis block, there has not a H-11 block.
+	if header == nil {
+		return true
+	}
+	// check whether H-11 block is a empty block.
+	if header.TxHash == types.EmptyRootHash {
+		parent := vm.bc.GetHeaderByHash(header.ParentHash)
+		if header.Root == parent.Root {
+			return true
```

Comment:
No need use if branch, actually if you use if branch, when not equal, you should use else branch to return false here, so recommend to `return header.Root == parent.Root ` directly.

Reply:
I think when header.Root == parent.Root, we should check whether the parent has been verified, if not, this empty block is not verified, if the parent has been verified already, this empty block is verified. So maybe here should be
`if header.Root != parent.Root {
    return false
}
`